File: pharmacy/controllers/stock_entry.py
import frappe
import pandas as pd
import json
from frappe import _
from erpnext.stock.get_item_details import get_conversion_factor

class NegativeStockError(frappe.ValidationError):
	pass

@frappe.whitelist()
def validate_items(doc,doc_event):
	pass
	# frappe.throw("kelani nbe7")
	# if len(doc.items)>0:
	# 	for item in doc.items:
	# 		batchno =frappe.db.get_list('Batch',filters={'disabled': 0,'item':item.item_code},fields=['name'],order_by='expiry_date Asc')
	# 		item.batch_no = batchno[0].name
    
def before_save(doc,method):
	validate_items_expiry_dates(doc)
	validate_items_qty(doc)

def validate_items_expiry_dates(doc):
	stock_entry_types = ["Material Issue", "Material Transfer"]
	stock_entry_type = doc.stock_entry_type
	if stock_entry_type not in stock_entry_types:
		return
	invalid_items = []
	for item in doc.items:
		if item.custom_using_date_expire == 1 and not item.item_expiry_date:
			invalid_items.append(f'{item.idx}')
	
	if invalid_items:
		invalid_items_idxs = ','.join(invalid_items)
		frappe.throw(
			title='Missing Fields',
    		msg=f'Rows: {invalid_items_idxs} &emsp;&emsp; <b>Source Item Expiry Date</b> IS Missing!',
		)

def validate_items_qty(doc):
	stock_entry_type = doc.stock_entry_type
	if stock_entry_type != "Material Transfer":
		return
	
	msg_list = []
	for item in doc.items:
		if item.custom_using_date_expire == 1 and item.qty > item.custom_expiry_date_qty:
			deficiency = round(item.qty - item.custom_expiry_date_qty, 2)
			msg = _("{0} units of {1} needed in {2} to complete this transaction.").format(
					frappe.bold(abs(deficiency)),
					frappe.get_desk_link("Item", item.item_code),
					frappe.get_desk_link("Warehouse", item.s_warehouse),
				)
			msg_list.append(msg)
	
	if msg_list:
		message = '<br>'.join(msg_list)
		frappe.throw(message, NegativeStockError, title=_("Insufficient Stock"))

@frappe.whitelist()
def get_expiry_date_qty(item_code = None, warehouse = None, expiry_date = None, uom = None):
	if item_code and warehouse and expiry_date and uom:
		item_details = frappe.db.sql("""
			SELECT warehouse,item_code,stock_uom,item_expiry_date, SUM(actual_qty) AS qty 
			FROM `tabStock Ledger Entry`
			WHERE is_cancelled = 0 AND item_code = %(item_code)s AND warehouse = %(warehouse)s AND item_expiry_date = %(item_expiry_date)s
			GROUP BY warehouse,item_code,stock_uom, item_expiry_date
			HAVING SUM(actual_qty) > 0
		""", values={
			'item_code': item_code,
			'warehouse': warehouse,
			'item_expiry_date': expiry_date
		},as_dict=1)

		expiry_date_qty = item_details[0].get('qty') if item_details else 0
		if expiry_date_qty:
			if item_details[0].get('stock_uom') != uom:
				conversion_factor = get_conversion_factor(item_code, uom).get('conversion_factor') or 1
				expiry_date_qty = expiry_date_qty / conversion_factor
		return expiry_date_qty
	else:
		return 0

@frappe.whitelist()
def get_expiry_date(item_code, warehouse):
	item_details = frappe.db.sql("""
		SELECT item_expiry_date
		FROM `tabStock Ledger Entry`
		WHERE is_cancelled = 0 AND item_code = %(item_code)s AND warehouse = %(warehouse)s AND item_expiry_date >= NOW()
		GROUP BY warehouse,item_code,stock_uom, item_expiry_date
		HAVING SUM(actual_qty) > 0
		ORDER BY item_expiry_date
	""", values={
		'item_code': item_code,
		'warehouse': warehouse
	},as_dict=1)

	return item_details[0].get('item_expiry_date') if item_details else ''

File: pharmacy/public/js/stock_entry.js
frappe.ui.form.on('Stock Entry', {
    setup(frm) {
        frm.set_query('uom','items', function(doc, cdt, cdn) {
           let row = locals[cdt][cdn];
            return {
                query: "pharmacy.controllers.stock_entry_item.get_item_uom",
                filters: {
                    'item_code': row.item_code
                }
            }
        })
    },
    validate(frm){
        if(frm.doc.stock_entry_type == 'Material Transfer'){
            frm.doc.items.forEach(item => {
                if(!item.s_warehouse){
                    frappe.model.set_value(item.doctype, item.name, 's_warehouse', frm.doc.from_warehouse)
                }
            });
        }
    }
})
    
frappe.ui.form.on('Stock Entry Detail', {
    item_code(frm,cdt,cdn){
        // frappe.model.set_value(cdt,cdn, 'qty', 0)
        // frappe.model.set_value(cdt,cdn, 'item_expiry_date', '')
        set_expiry_date(frm,cdt,cdn)
        set_expiry_date_qty(frm,cdt,cdn)
    },
    s_warehouse(frm,cdt,cdn){
        set_expiry_date(frm,cdt,cdn)
        set_expiry_date_qty(frm,cdt,cdn)
    },
    item_expiry_date(frm,cdt,cdn){
        set_expiry_date_qty(frm,cdt,cdn)
    },
    uom(frm,cdt,cdn){
        set_expiry_date_qty(frm,cdt,cdn)
    },
    qty(frm,cdt,cdn){
        if(frm.doc.stock_entry_type == 'Material Transfer'){
            set_expiry_date_qty(frm,cdt,cdn)
            setTimeout(()=>{
                let row = locals[cdt][cdn]
                if(row.custom_using_date_expire && row.qty > row.custom_expiry_date_qty){
                    frappe.model.set_value(cdt,cdn, 'qty', 0)
                    frappe.msgprint("Qty Must Be Less Than Or Equal To Expiry Date Qty!")
                }
            }, 500)
        }
    },
})

function set_expiry_date_qty(frm,cdt,cdn){
    let row = locals[cdt][cdn]
    if(row.custom_using_date_expire){
        frappe.call({
            method: 'pharmacy.controllers.stock_entry.get_expiry_date_qty',
            args:{
                item_code: row.item_code,
                warehouse: row.s_warehouse,
                expiry_date: row.item_expiry_date,
                uom: row.uom
            },
            callback:(r)=>{
                frappe.model.set_value(cdt,cdn, 'custom_expiry_date_qty', r.message)
            }
        })
    }
}

function set_expiry_date(frm,cdt,cdn){
    let row = locals[cdt][cdn]
    if(frm.doc.stock_entry_type == 'Material Transfer' && row.item_code && row.s_warehouse){
        frappe.call({
            method: 'pharmacy.controllers.stock_entry.get_expiry_date',
            args:{
                item_code: row.item_code,
                warehouse: row.s_warehouse
            },
            callback:(r)=>{
                frappe.model.set_value(cdt,cdn, 'item_expiry_date', r.message)
            }
        })
    }
}